File: src/test/python/CommerceTest.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.common.keys import Keys


class TestCommerce():
    
    USERNAME = "standard_user"
    LOCKEDUSERNAME = "locked_out_user"
    GLITCHUSERNAME = "performance_glitch_user"
    PASSWORD = "secret_sauce"
    
    driver = None
    wait = None
    actions = None
    
    def set_up(self, url):
        dimension = (1920, 1080)
        options = webdriver.ChromeOptions()
        options.add_argument("--headless=new")
        self.driver = webdriver.Chrome(options=options)
        self.driver.implicitly_wait(4)  # seconds
        self.driver.set_window_size(*dimension)
        self.wait = WebDriverWait(self.driver, 10)  # seconds
        self.driver.get(url)

        self.actions = ActionChains(self.driver)
    
    def log_in(self, name, password):
        self.driver.find_element(By.ID, "user-name").send_keys(name)
        self.driver.find_element(By.ID, "password").send_keys(password)
        self.driver.find_element(By.ID, "login-button").click()

        elements = self.driver.find_elements(By.ID, "inventory_container")
        assert len(elements) > 0

    def log_out(self):
        self.driver.find_element(By.ID, "react-burger-menu-btn").click()
        self.driver.find_element(By.ID, "logout_sidebar_link").click()

        elements = self.driver.find_elements(By.CLASS_NAME, "login_container")
        assert len(elements) > 0
        
    def tearDown(self):
        self.log_out()
        self.driver.quit()

    def buy_backpack(self):

        self.set_up("https://www.saucedemo.com")
        self.log_in(self.USERNAME, self.PASSWORD)

        # Click "Add to cart" for the backpack item
        self.driver.find_element(By.ID, "add-to-cart-sauce-labs-backpack").click()
        # Click on the cart icon
        self.driver.find_element(By.ID, "shopping_cart_container").click()

        # Explicit wait until the page is loaded.
        cart_page_element = self.wait.until(EC.visibility_of_element_located((By.CLASS_NAME, "cart_list")))

        # Assert that the item name is correctly displayed on the
        # cart page.
        cart_item_name = self.driver.find_element(By.CLASS_NAME, "inventory_item_name").text
        assert cart_item_name == "Sauce Labs Backpack"

        # Click on the checkout button
        self.driver.find_element(By.ID, "checkout").click()
        # Wait until the checkout page is displayed
        checkout_page_element = self.wait.until(EC.visibility_of_element_located((By.ID, "checkout_info_container")))

        # Fill in the credentials
        self.driver.find_element(By.ID, "first-name").send_keys("John")
        self.driver.find_element(By.ID, "last-name").send_keys("Doe")
        self.driver.find_element(By.ID, "postal-code").send_keys("10001")
        self.driver.find_element(By.ID, "continue").click()

        # Assert that the checkout item name is correct
        checkout_item_name = self.driver.find_element(By.CLASS_NAME, "inventory_item_name").text
        assert checkout_item_name == "Sauce Labs Backpack"

        # Click on the finish button
        self.driver.find_element(By.ID, "finish").click()

        # Assert that the success page is visible
        success_is_displayed = self.driver.find_element(By.ID, "checkout_complete_container").is_displayed()
        assert success_is_displayed

        self.tearDown()      
        
    def test_assert_item_name_on_item_info_page(self):
        self.set_up("https://www.saucedemo.com")
        self.log_in(self.USERNAME, self.PASSWORD)

        # go to the item
        self.driver.find_element(By.ID, "item_4_title_link").click()

        # Check that texts exist on the site
        # Sauce Labs Backpack
        assert self.driver.find_element(By.CLASS_NAME, "inventory_details_name").text == "Sauce Labs Backpack"
        
        expected_desc = "carry.allTheThings() with the sleek, streamlined Sly Pack that melds uncompromising style with unequaled laptop and tablet protection."
        assert self.driver.find_element(By.CLASS_NAME, "inventory_details_desc").text == expected_desc
        
        self.tearDown()

    def test_move_to_linkedin(self):
        self.set_up("https://www.saucedemo.com")
        self.log_in(self.USERNAME, self.PASSWORD)

        # Scroll down
        self.actions.move_to_element(self.driver.find_element(By.CLASS_NAME, "social_linkedin")).perform()

        # Go to Linkedin page
        self.driver.find_element(By.CLASS_NAME, "social_linkedin").click()

        # Verify you are on the Linkedin page
        tab_list = self.driver.window_handles
        self.driver.switch_to.window(tab_list[1])

        assert self.driver.current_url == "https://www.linkedin.com/company/sauce-labs/"
        self.driver.close()
        self.driver.switch_to.window(tab_list[0])

        self.tearDown()    
        
    def test_assert_sorting(self):
        self.set_up("https://www.saucedemo.com")
        self.log_in(self.USERNAME, self.PASSWORD)

        # Click the filters box
        self.driver.find_element(By.CLASS_NAME, "select_container").click()
        # click Z-A
        self.driver.find_element(By.CSS_SELECTOR, "option[value='za']").click()
        # check that the name at the top is "Test.allTheThings() T-Shirt (Red)
        assert self.driver.find_element(By.CSS_SELECTOR, ".inventory_item_name").text == "Test.allTheThings() T-Shirt (Red)"

        # click the filters box
        self.driver.find_element(By.CLASS_NAME, "select_container").click()
        # click A-Z 
        self.driver.find_element(By.CSS_SELECTOR, "option[value='az']").click()
        # check that the name at the top is "Sauce Labs Backpack"
        assert self.driver.find_element(By.CSS_SELECTOR, ".inventory_item_name").text == "Sauce Labs Backpack"

        # click the filters box
        self.driver.find_element(By.CLASS_NAME, "select_container").click()
        # click low-High
        self.driver.find_element(By.CSS_SELECTOR, "option[value='hilo']").click()
        # check that the name at the top is "Sauce Labs Onesie"
        assert self.driver.find_element(By.CSS_SELECTOR, ".inventory_item_name").text == "Sauce Labs Fleece Jacket"

        # click the filters box
        self.driver.find_element(By.CLASS_NAME, "select_container").click()
        # click high-Low
        self.driver.find_element(By.CSS_SELECTOR, "option[value='lohi']").click()
        # check that the name at the top is "Sauce Labs Fleece Jacket"
        assert self.driver.find_element(By.CSS_SELECTOR, ".inventory_item_name").text == "Sauce Labs Onesie"
        self.tearDown()

    def test_assert_shopping_button_states(self):
        self.set_up("https://www.saucedemo.com")
        self.log_in(self.USERNAME, self.PASSWORD)

        # Click all add-to-cart buttons
        buttons_add = self.driver.find_elements(By.CSS_SELECTOR, "[id^=add-to-cart]")
        for button in buttons_add:
            self.actions.move_to_element(button).perform()
            button.click()

        # check that the cart button says 6
        assert self.driver.find_element(By.CLASS_NAME, "shopping_cart_badge").text == "6"

        # Click all remove buttons
        buttons_remove = self.driver.find_elements(By.CSS_SELECTOR, "[id^=remove]")
        for button in buttons_remove:
            self.actions.move_to_element(button).perform()
            button.click()

        # check that the cart button says nothing
        assert len(self.driver.find_elements(By.CLASS_NAME, "shopping_cart_badge")) < 1
        self.tearDown()

    def test_assert_user_log_in_cache_permanency(self):
        self.set_up("https://www.saucedemo.com")
        self.log_in(self.USERNAME, self.PASSWORD)
        
        # Add items to cart
        self.actions.move_to_element(self.driver.find_element(By.ID, "add-to-cart-sauce-labs-backpack")).perform()
        self.driver.find_element(By.ID, "add-to-cart-sauce-labs-backpack").click()

        self.actions.move_to_element(self.driver.find_element(By.ID, "add-to-cart-sauce-labs-bike-light")).perform()
        self.driver.find_element(By.ID, "add-to-cart-sauce-labs-bike-light").click()

        self.log_out()
        # Check that items are still in the cart
        self.log_in(self.USERNAME, self.PASSWORD)

        assert self.driver.find_element(By.CLASS_NAME, "shopping_cart_badge").text == "2"
        self.tearDown()

    def test_assert_app_reset_functionality(self):  # flaky test
        self.set_up("https://www.saucedemo.com")
        self.log_in(self.USERNAME, self.PASSWORD)

        # Select all items
        buttons_add = self.driver.find_elements(By.CSS_SELECTOR, "[id^=add-to-cart]")
        for button in buttons_add:
            self.actions.move_to_element(button).perform()
            button.click()

        # Click on sidebar
        self.driver.find_element(By.ID, "react-burger-menu-btn").click()
        # Click on Reset App State
        self.driver.find_element(By.ID, "reset_sidebar_link").click()
        # Click on the close sidebar button using JS.
        # This was necessary because normal click was not waiting 
        # for the CSS animation to be over, resulting in an
        # element click interception.
        self.driver.execute_script("arguments[0].click();", self.driver.find_element(By.ID, "react-burger-cross-btn"))

        # Check that the app reset but that there is a bug with "Remove" still staying instead of resetting to Add Cart
        assert len(self.driver.find_elements(By.CLASS_NAME, "shopping_cart_badge")) < 1

        assert self.driver.find_element(By.ID, "remove-sauce-labs-backpack").is_displayed()
        assert self.driver.find_element(By.ID, "remove-sauce-labs-bike-light").is_displayed()
        assert self.driver.find_element(By.ID, "remove-sauce-labs-bolt-t-shirt").is_displayed()
        assert self.driver.find_element(By.ID, "remove-sauce-labs-fleece-jacket").is_displayed()
        assert self.driver.find_element(By.ID, "remove-sauce-labs-onesie").is_displayed()
        assert self.driver.find_element(By.ID, "remove-test.allthethings()-t-shirt-(red)").is_displayed()

        self.tearDown()
        
    def test_assert_error_accessing_without_log_in(self):
        self.set_up("https://www.saucedemo.com/cart.html")  # Direct link provided here

        # Assert that the error is the correct one
        expected_error = "Epic sadface: You can only access '/cart.html' when you are logged in."
        assert expected_error == self.driver.find_element(By.CLASS_NAME, "error-message-container").text

        self.log_in(self.USERNAME, self.PASSWORD)

        # Now try logging in and then jumping to a link
        self.driver.get("https://www.saucedemo.com/cart.html")

        # Assert that it worked when you were logged in
        assert self.driver.find_element(By.ID, "cart_contents_container").is_displayed()
        self.tearDown()

    def test_assert_locked_out_user(self):
        self.set_up("https://www.saucedemo.com")

        # Attempt to log in with locked out user
        self.driver.find_element(By.ID, "user-name").send_keys(self.LOCKEDUSERNAME)
        self.driver.find_element(By.ID, "password").send_keys(self.PASSWORD)
        self.driver.find_element(By.ID, "login-button").click()

        # Clear the input fields
        self.driver.find_element(By.ID, "user-name").send_keys(Keys.CONTROL + "a", Keys.BACK_SPACE)
        self.driver.find_element(By.ID, "password").send_keys(Keys.CONTROL + "a", Keys.BACK_SPACE)

        # Assert the error
        expected_error = "Epic sadface: Sorry, this user has been locked out."
        actual_error = self.driver.find_element(By.CLASS_NAME, "error-message-container").text
        assert expected_error == actual_error

        # Attempt to log in with the normal user
        self.log_in(self.USERNAME, self.PASSWORD)
        
        self.tearDown()
        
    def test_glitched_User_Experience(self):

        self.set_up("https://www.saucedemo.com")
        self.log_in(self.USERNAME, self.PASSWORD)

        self.driver.find_element(By.ID, "add-to-cart-sauce-labs-backpack").click()
        self.driver.find_element(By.ID, "shopping_cart_container").click()

        cart_page_element = self.wait.until(EC.visibility_of_element_located((By.CLASS_NAME, "cart_list")))
        assert self.driver.find_element(By.CLASS_NAME, "inventory_item_name").text == "Sauce Labs Backpack"

        self.driver.find_element(By.ID, "checkout").click()
        checkout_page_element = self.wait.until(EC.visibility_of_element_located((By.ID, "checkout_info_container")))

        self.driver.find_element(By.ID, "first-name").send_keys("John")
        self.driver.find_element(By.ID, "last-name").send_keys("Doe")
        self.driver.find_element(By.ID, "postal-code").send_keys("10001")
        self.driver.find_element(By.ID, "continue").click()

        checkout_item_name = self.driver.find_element(By.CLASS_NAME, "inventory_item_name").text
        assert checkout_item_name == "Sauce Labs Backpack"

        self.driver.find_element(By.ID, "finish").click()
        success_is_displayed = self.driver.find_element(By.ID, "checkout_complete_container").is_displayed()
        assert success_is_displayed

        self.tearDown()

File: src/test/Ruby/CommerceTest.rb
require 'selenium-webdriver'
require 'test/unit'

class CommerceTest < Test::Unit::TestCase
  USERNAME = 'standard_user'.freeze
  LOCKEDUSERNAME = 'locked_out_user'.freeze
  GLITCHUSERNAME = 'performance_glitch_user'.freeze
  PASSWORD = 'secret_sauce'.freeze

  def setup_driver(url)
    options = Selenium::WebDriver::Chrome::Options.new
    options.add_argument('--headless=new')
    @driver = Selenium::WebDriver.for :chrome, options: options
    @driver.manage.timeouts.implicit_wait = 4 # seconds
    @driver.manage.window.resize_to(1920, 1080)
    @wait = Selenium::WebDriver::Wait.new(timeout: 10) # seconds
    @driver.navigate.to url

    @actions = @driver.action
  end

  def log_in(username, password)
    @driver.find_element(id: 'user-name').send_keys(username)
    @driver.find_element(id: 'password').send_keys(password)
    @driver.find_element(id: 'login-button').click

    elements = @driver.find_elements(id: 'inventory_container')
    assert_true(elements.size > 0)
  end

  def log_out
    @driver.find_element(id: 'react-burger-menu-btn').click
    @driver.find_element(id: 'logout_sidebar_link').click

    elements = @driver.find_elements(class: 'login_container')
    assert_true(elements.size > 0)
  end

  def teardown
    log_out
    @driver.quit
  end

  def buy_backpack
    setup_driver('https://www.saucedemo.com')
    log_in(USERNAME, PASSWORD)

    # Click "Add to cart" for the backpack item
    @driver.find_element(id: 'add-to-cart-sauce-labs-backpack').click
    # Click on the cart icon
    @driver.find_element(id: 'shopping_cart_container').click

    # Explicit wait until the page is loaded.
    cart_page_element = @driver.find_element(class: 'cart_list')
    @wait.until { cart_page_element.displayed? }

    # Assert that the item name is correctly displayed on the
    # cart page.
    cart_item_name = @driver.find_element(class: 'inventory_item_name').text
    assert_equal(cart_item_name, 'Sauce Labs Backpack')

    # Click on the checkout button
    @driver.find_element(id: 'checkout').click

    # Wait until the checkout page is displayed
    checkout_page_element = @driver.find_element(id: 'checkout_info_container')
    @wait.until { checkout_page_element.displayed? }

    # Fill in the credentials
    @driver.find_element(id: 'first-name').send_keys('John')
    @driver.find_element(id: 'last-name').send_keys('Doe')
    @driver.find_element(id: 'postal-code').send_keys('10001')
    @driver.find_element(id: 'continue').click

    # Assert that the checkout item name is correct
    checkout_item_name = @driver.find_element(class: 'inventory_item_name').text
    assert_equal(checkout_item_name, 'Sauce Labs Backpack')

    # Click on the finish button
    @driver.find_element(id: 'finish').click

    # Assert that the success page is visible
    success_is_displayed = @driver.find_element(id: 'checkout_complete_container').displayed?
    assert_true(success_is_displayed)
  end

  def test_assert_item_name_on_item_info_page
    setup_driver('https://www.saucedemo.com')
    log_in(USERNAME, PASSWORD)

    # go to the item
    @driver.find_element(id: 'item_4_title_link').click

    # check that texts exist on the site
    assert_equal(@driver.find_element(class: 'inventory_details_name').text, "Sauce Labs Backpack")
    expected = "carry.allTheThings() with the sleek, streamlined Sly Pack that melds uncompromising style with unequaled laptop and tablet protection."
    assert_equal(@driver.find_element(class: 'inventory_details_desc').text, expected)

  end

  def test_move_to_linkedin
    setup_driver('https://www.saucedemo.com')
    log_in(USERNAME, PASSWORD)
  
    # Scroll down
    linkedin_element = @driver.find_element(class: 'social_linkedin')
    @actions.move_to(linkedin_element).perform
    # Go to Linkedin page
    linkedin_element.click
  
    # Verify you are on the Linkedin page
    tab_list = @driver.window_handles
    @driver.switch_to.window(tab_list[1])
  
    assert_equal('https://www.linkedin.com/company/sauce-labs/', @driver.current_url)
    @driver.close
    @driver.switch_to.window(tab_list[0])
  end
  
  def test_assert_sorting
    setup_driver('https://www.saucedemo.com')
    log_in(USERNAME, PASSWORD)
  
    # Click the filters box and select Z-A
    @driver.find_element(class: 'select_container').click
    @driver.find_element(css: "option[value='za']").click
    # Check that the name at the top is "Test.allTheThings() T-Shirt (Red)"
    assert_equal('Test.allTheThings() T-Shirt (Red)', @driver.find_element(css: '.inventory_item_name').text)
  
    # Click the filters box and select A-Z
    @driver.find_element(class: 'select_container').click
    @driver.find_element(css: "option[value='az']").click
    # Check that the name at the top is "Sauce Labs Backpack"
    assert_equal('Sauce Labs Backpack', @driver.find_element(css: '.inventory_item_name').text)
  
    # Click the filters box and select Low-High
    @driver.find_element(class: 'select_container').click
    @driver.find_element(css: "option[value='hilo']").click
    # Check that the name at the top is "Sauce Labs Onesie"
    assert_equal('Sauce Labs Fleece Jacket', @driver.find_element(css: '.inventory_item_name').text)
  
    # Click the filters box and select High-Low
    @driver.find_element(class: 'select_container').click
    @driver.find_element(css: "option[value='lohi']").click
    # Check that the name at the top is "Sauce Labs Fleece Jacket"
    assert_equal('Sauce Labs Onesie', @driver.find_element(css: '.inventory_item_name').text)
  end

  def test_assert_shopping_button_states
    setup_driver('https://www.saucedemo.com')
    log_in(USERNAME, PASSWORD)
  
    # Add items to cart
    @driver.find_elements(css: "[id^=add-to-cart]").each do |button|
      @actions.move_to(button).perform
      button.click
    end
  
    # Check that the cart button says 6
    assert_equal('6', @driver.find_element(class: 'shopping_cart_badge').text)
  
    # Remove items from cart
    @driver.find_elements(css: "[id^=remove]").each do |button|
      @actions.move_to(button).perform
      button.click
    end
  
    # Check that the cart button says nothing
    assert_true(@driver.find_elements(class: 'shopping_cart_badge').size < 1)
  end
  
  def test_assert_user_log_in_cache_permanency
    setup_driver('https://www.saucedemo.com')
    log_in(USERNAME, PASSWORD)
  
    # Add items to cart
    backpack_button = @driver.find_element(id: 'add-to-cart-sauce-labs-backpack')
    @actions.move_to(backpack_button).perform
    backpack_button.click
  
    bike_light_button = @driver.find_element(id: 'add-to-cart-sauce-labs-bike-light')
    @actions.move_to(bike_light_button).perform
    bike_light_button.click
  
    log_out
  
    # Check that items are still in the cart
    log_in(USERNAME, PASSWORD)
    assert_equal('2', @driver.find_element(class: 'shopping_cart_badge').text)
  end

  def test_assert_app_reset_functionality  # flaky test
    setup_driver('https://www.saucedemo.com')
    log_in(USERNAME, PASSWORD)
  
    # Select all items
    @driver.find_elements(css: "[id^=add-to-cart]").each do |button|
      @actions.move_to(button).perform
      button.click
    end
  
    # Click on sidebar
    @driver.find_element(id: 'react-burger-menu-btn').click
    # Click on Reset App State
    @driver.find_element(id: 'reset_sidebar_link').click
    # Click on the close sidebar button using JS.
    # This was necessary because normal click was not waiting 
    # for the CSS animation to be over, resulting in an
    # element click interception.
    @driver.execute_script("arguments[0].click();", @driver.find_element(id: 'react-burger-cross-btn'))
  
    # Check that the app reset but that there is a bug with "Remove" still staying instead of resetting to Add Cart
    assert_true(@driver.find_elements(class: 'shopping_cart_badge').size < 1)
  
    # Verify the visibility of 'remove' buttons
    assert_true(@driver.find_element(id: 'remove-sauce-labs-backpack').displayed?)
    assert_true(@driver.find_element(id: 'remove-sauce-labs-bike-light').displayed?)
    assert_true(@driver.find_element(id: 'remove-sauce-labs-bolt-t-shirt').displayed?)
    assert_true(@driver.find_element(id: 'remove-sauce-labs-fleece-jacket').displayed?)
    assert_true(@driver.find_element(id: 'remove-sauce-labs-onesie').displayed?)
    assert_true(@driver.find_element(id: 'remove-test.allthethings()-t-shirt-(red)').displayed?)
  end
  
  def test_assert_error_accessing_without_log_in
    setup_driver('https://www.saucedemo.com/cart.html')  # write the direct link here
  
    # Assert that the error is the correct one
    expected = 'Epic sadface: You can only access \'/cart.html\' when you are logged in.'
    assert_equal(expected, @driver.find_element(class: 'error-message-container').text)
  
    log_in(USERNAME, PASSWORD)
  
    # Now try logging in and then jumping to a link
    @driver.navigate.to('https://www.saucedemo.com/cart.html')
  
    # Assert that it worked when you were logged in
    assert_true(@driver.find_element(id: 'cart_contents_container').displayed?)
  end

  def test_assert_locked_out_user
    setup_driver('https://www.saucedemo.com')
    
    # Enter LOCKEDUSERNAME and PASSWORD and try to log in
    @driver.find_element(id: 'user-name').send_keys(LOCKEDUSERNAME)
    @driver.find_element(id: 'password').send_keys(PASSWORD)
    @driver.find_element(id: 'login-button').click
    
    # Clear the input fields
    @driver.find_element(id: 'user-name').send_keys(:control, 'a', :backspace)
    @driver.find_element(id: 'password').send_keys(:control, 'a', :backspace)
    
    # Assert the error
    expected = 'Epic sadface: Sorry, this user has been locked out.'
    assert_equal(expected, @driver.find_element(class: 'error-message-container').text)
    
    # Attempt to log in with the normal user
    log_in(USERNAME, PASSWORD)
  end

  def test_glitched_user_experience
    setup_driver('https://www.saucedemo.com')
    log_in(GLITCHUSERNAME, PASSWORD)

    @driver.find_element(id: 'add-to-cart-sauce-labs-backpack').click
    @driver.find_element(id: 'shopping_cart_container').click

    cart_page_element = @driver.find_element(class: 'cart_list')
    @wait.until { cart_page_element.displayed? }

    cart_item_name = @driver.find_element(class: 'inventory_item_name').text
    assert_equal(cart_item_name, 'Sauce Labs Backpack')

    @driver.find_element(id: 'checkout').click
    checkout_page_element = @driver.find_element(id: 'checkout_info_container')
    @wait.until { checkout_page_element.displayed? }

    @driver.find_element(id: 'first-name').send_keys('John')
    @driver.find_element(id: 'last-name').send_keys('Doe')
    @driver.find_element(id: 'postal-code').send_keys('10001')
    @driver.find_element(id: 'continue').click

    checkout_item_name = @driver.find_element(class: 'inventory_item_name').text
    assert_equal(checkout_item_name, 'Sauce Labs Backpack')

    @driver.find_element(id: 'finish').click
    success_is_displayed = @driver.find_element(id: 'checkout_complete_container').displayed?
    assert_true(success_is_displayed)
  end
  
end
